### PR TITLE
Add units::angle and port various code to use it

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
SUMMARY: Infrastructure "Add typesafe angle type"

#### Purpose of change
For general type safety but also to avoid confusion between different measures of angle (radians, degrees, and arcmin are all used in this codebase).

#### Describe the solution
Add a typesafe angle type.

Convert tileray to use the new type, and deal with the fallout from other things that depend on tileray.

Most of the changes relate to changing vehicle facing to the new type.

Note that this changes vehicle facing from an integer to a floating point type, which could cause some behavioural change.  However, the angles involved are still serialized as integer degrees.

#### Describe alternatives you've considered
Could have stuck with integer degrees or integer arcmin rather than floating point radians as the default angle type.  But decided to stick with radians to reduce conversions when calling library functions, and because it just felt like the best option.

#### Testing
Added some unit tests.

Experimented with driving a vehicle in-game; that did find some bugs which I fixed.

Checked that saving and loading vehicles works.  That also found bugs which I fixed.

#### Additional context
Inspired in particular because I got confused between degrees and radians in another feature branch.

### CODE REVIEW

File: src/savegame_json.cpp [hunk 1]
```diff
@@ -2783,7 +2785,9 @@ void vehicle::deserialize( JsonIn &jsin )
     data.read( "om_id", om_id );
     data.read( "faceDir", fdir );
     data.read( "moveDir", mdir );
-    data.read( "turn_dir", turn_dir );
+    int turn_dir_int;
+    data.read( "turn_dir", turn_dir_int );
+    turn_dir = units::from_degrees( turn_dir_int );
```
Comment (KorGgenT) [inline src/savegame_json.cpp:2790]: it would be nice to add a serialize/deserialize function for this like the current ones in units.cpp

**Comment by KorGgenT** [comment]
I'm glad to see this. i read through the parts of the code related to magic and it looks good. thanks for that small refactor, too. I only skimmed the rest of it.

File: src/savegame_json.cpp (hunk 1 above)
Comment (jbytheway) [reply src/savegame_json.cpp:2790]: (De)serialization of angles actually already works, but for backwards compatibility we still need to support reading integer degrees here.  The best option might have been to add support for both, but I decided to punt on that for the time being.

**Comment by jbytheway** [comment]
One other thing I forgot to mention in the original post: I tried also porting weapon dispersion to the new `angle` type, but the code dealing with dispersion and recoil is super confusing, and I sometimes couldn't figure out what units it was using, so I backed out those changes.

**Comment by jbytheway** [comment]
I dealt with the LGTM issue.

File: src/veh_interact.cpp [hunk 2]
```diff
@@ -3003,15 +3003,7 @@ void veh_interact::complete_vehicle( player &p )
...
-                int dir = static_cast<int>( atan2( static_cast<float>( delta.y ),
-                                                   static_cast<float>( delta.x ) ) * 180.0 / M_PI );
-                dir -= veh->face.dir();
-                while( dir < 0 ) {
-                    dir += 360;
-                }
-                while( dir > 360 ) {
-                    dir -= 360;
-                }
+                units::angle dir = normalize( atan2( delta ) - veh->face.dir() );
```
Comment (mlangsdorf) [inline src/veh_interact.cpp:3006]: Do you ever get tired of making the game better in every possible way?  This is great stuff.

**Review by mlangsdorf — APPROVED** [review]
I reviewed the changes to the vehicle code.  90% straightforward refactor, 10% simplification and implementation of common functions.  Excellent stuff.

**Comment by kevingranade** [comment]
You picked up some breakage from another PR, should be trivial to fix.

**Comment by jbytheway** [comment]
I believe that's fixed now.

**Pull request merged**